An error occurred while executing the following cell:
------------------
try:
    import sys, os
except ModuleNotFoundError:
    %pip install -qq sys,
    import sys, os
try:
    import torch
except ModuleNotFoundError:
    %pip install -qq torch
    import torch
try:
    from torchvision.datasets import CelebA, MNIST
except ModuleNotFoundError:
    %pip install -qq torchvision
    from torchvision.datasets import CelebA, MNIST
import torchvision.transforms as transforms

try:
    from pytorch_lightning import LightningDataModule, LightningModule, Trainer
except ModuleNotFoundError:
    %pip install -qq pytorch_lightning
    from pytorch_lightning import LightningDataModule, LightningModule, Trainer
from torch.utils.data import DataLoader, random_split, SequentialSampler, RandomSampler
import numpy as np
from matplotlib import pyplot as plt

try:
    from imageio import imwrite
except ModuleNotFoundError:
    %pip install -qq imageio
    from imageio import imwrite
from packaging import version

try:
    from tqdm import tqdm
except ModuleNotFoundError:
    %pip install -qq tqdm
    from tqdm import tqdm
try:
    from data import CelebADataset, CelebADataModule
except ModuleNotFoundError:
    %pip install -qq data
    from data import CelebADataset, CelebADataModule
try:
    from mfa_celeba_helpers import *
except ModuleNotFoundError:
    %pip install -qq mfa_celeba_helpers
    from mfa_celeba_helpers import *
from IPython.display import Image


def main(argv):
    assert version.parse(torch.__version__) >= version.parse("1.2.0")

    dataset = argv[1] if len(argv) == 2 else "celeba"
    print("Preparing dataset and parameters for", dataset, "...")

    if dataset == "celeba":
        image_shape = [64, 64, 3]  # The input image shape
        n_components = 300  # Number of components in the mixture model
        n_factors = 10  # Number of factors - the latent dimension (same for all components)
        batch_size = 1000  # The EM batch size
        num_iterations = 30  # Number of EM iterations (=epochs)
        feature_sampling = 0.2  # For faster responsibilities calculation, randomly sample the coordinates (or False)
        mfa_sgd_epochs = 0  # Perform additional training with diagonal (per-pixel) covariance, using SGD
        init_method = "rnd_samples"  # Initialize each component from few random samples using PPCA
        trans = transforms.Compose(
            [
                CropTransform((25, 50, 25 + 128, 50 + 128)),
                transforms.Resize(image_shape[0]),
                transforms.ToTensor(),
                ReshapeTransform([-1]),
            ]
        )
        train_set = CelebADataset(root="./data", split="train", transform=trans, download=True)
        test_set = CelebADataset(root="./data", split="test", transform=trans, download=True)

    elif dataset == "mnist":
        image_shape = [28, 28]  # The input image shape
        n_components = 50  # Number of components in the mixture model
        n_factors = 6  # Number of factors - the latent dimension (same for all components)
        batch_size = 1000  # The EM batch size
        num_iterations = 1  # Number of EM iterations (=epochs)
        feature_sampling = False  # For faster responsibilities calculation, randomly sample the coordinates (or False)
        mfa_sgd_epochs = 0  # Perform additional training with diagonal (per-pixel) covariance, using SGD
        init_method = "kmeans"  # Initialize by using k-means clustering
        trans = transforms.Compose([transforms.ToTensor(), ReshapeTransform([-1])])
        train_set = MNIST(root="./data", train=True, transform=trans, download=True)
        test_set = MNIST(root="./data", train=False, transform=trans, download=True)
    else:
        assert False, "Unknown dataset: " + dataset
------------------

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
/tmp/ipykernel_5481/4106335031.py in <module>
     11 try:
---> 12     from torchvision.datasets import CelebA, MNIST
     13 except ModuleNotFoundError:

ModuleNotFoundError: No module named 'torchvision'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_5481/4106335031.py in <module>
     13 except ModuleNotFoundError:
     14     get_ipython().run_line_magic('pip', 'install -qq torchvision')
---> 15     from torchvision.datasets import CelebA, MNIST
     16 import torchvision.transforms as transforms
     17 

~/miniconda3/envs/py37/lib/python3.7/site-packages/torchvision/__init__.py in <module>
      3 
      4 import torch
----> 5 from torchvision import datasets
      6 from torchvision import io
      7 from torchvision import models

~/miniconda3/envs/py37/lib/python3.7/site-packages/torchvision/datasets/__init__.py in <module>
----> 1 from ._optical_flow import KittiFlow, Sintel, FlyingChairs, FlyingThings3D, HD1K
      2 from .caltech import Caltech101, Caltech256
      3 from .celeba import CelebA
      4 from .cifar import CIFAR10, CIFAR100
      5 from .cityscapes import Cityscapes

~/miniconda3/envs/py37/lib/python3.7/site-packages/torchvision/datasets/_optical_flow.py in <module>
     10 from PIL import Image
     11 
---> 12 from ..io.image import _read_png_16
     13 from .utils import verify_str_arg
     14 from .vision import VisionDataset

~/miniconda3/envs/py37/lib/python3.7/site-packages/torchvision/io/__init__.py in <module>
      3 import torch
      4 
----> 5 from ..utils import _log_api_usage_once
      6 
      7 try:

~/miniconda3/envs/py37/lib/python3.7/site-packages/torchvision/utils.py in <module>
     19 
     20 
---> 21 @torch.no_grad()
     22 def make_grid(
     23     tensor: Union[torch.Tensor, List[torch.Tensor]],

AttributeError: module 'torch' has no attribute 'no_grad'
AttributeError: module 'torch' has no attribute 'no_grad'
